tests/src/python/test_qgsserver_offline_editing.py

```diff
@@ -0,0 +1,275 @@
+        assert layer.isValid()
+        return layer
+
+    @classmethod
+    def _getWFSLayer(cls, type_name, layer_name=None):
+        """
+        WFS layer factory
+        """
+        if layer_name is None:
+            layer_name = 'wfs_' + type_name
+        parms = {
+            'srsname': 'EPSG:4326',
+            'typename': type_name,
+            'url': 'http://127.0.0.1:%s/?map=%s' % (cls.port,
+                                                    cls.project_path),
+            'version': 'auto',
+            'table': '',
+            #'sql': '',
+        }
+        uri = ' '.join([("%s='%s'" % (k, v)) for k, v in parms.iteritems()])
+        wfs_layer = QgsVectorLayer(uri, layer_name, 'WFS')
+        assert wfs_layer.isValid()
+        return wfs_layer
+
+    @classmethod
+    def _getFeatureByAttribute(cls, layer, attr_name, attr_value):
+        """
+        Find the feature and return it, raise exception if not found
+        """
+        request = QgsFeatureRequest(QgsExpression("%s=%s" % (attr_name,
+                                                             attr_value)))
+        try:
+            return layer.dataProvider().getFeatures(request).next()
+        except StopIteration:
+            raise Exception("Wrong attributes in WFS layer %s" %
+                            layer.name())
+
+    def test_offlineConversion(self):
+        # goes offline
+        ol = QgsOfflineEditing()
+        wfs_layer = self.registry.mapLayers().values()[0]
+        self.assertTrue(wfs_layer.hasGeometryType())
+        # Check we have 3 features
+        self.assertEqual(len([f for f in wfs_layer.getFeatures()]), len(TEST_FEATURES))
+        self.assertTrue(ol.convertToOfflineProject(self.temp_path, 'offlineDbFile.sqlite', [wfs_layer.id()]))
+        offline_layer = self.registry.mapLayers().values()[0]
+        self.assertTrue(offline_layer.hasGeometryType())
+        self.assertTrue(offline_layer.isValid())
+        self.assertTrue(offline_layer.name().find('(offline)') > -1)
+        self.assertEqual(len([f for f in offline_layer.getFeatures()]), len(TEST_FEATURES))
+        # Edit feature 2
+        feat2 = self._getFeatureByAttribute(offline_layer, 'name', "'name 2'")
+        self.assertTrue(offline_layer.startEditing())
+        self.assertTrue(offline_layer.changeAttributeValue(feat2.id(), offline_layer.fieldNameIndex('name'), 'name 2 edited'))
+        self.assertTrue(offline_layer.changeGeometry(feat2.id(), QgsGeometry.fromPoint(QgsPoint(33.0, 60.0))))
+        self.assertTrue(offline_layer.commitChanges())
+        feat2 = self._getFeatureByAttribute(offline_layer, 'name', "'name 2 edited'")
+        self.assertTrue(ol.isOfflineProject())
+        self.assertTrue(self.project.readEntry("OfflineEditingPlugin", "/OfflineDbPath")[1])
+        # Sync
+        ol.synchronize()
+        # Does anybody knows why the sleep is needed? Is that a threaded WFS consequence?
+        sleep(1)
```

Comment:
Hum, no definite idea to explain the need for the sleep. I can't think to a reason on the client side (but I can be wrong). A potential explanation would be that the server acknowledges the WFS-T request to the client, but would commit the changes in a separate thread, but I doubt it is architectured like that (I haven't checked).